assets/shiny-server.js

```diff
@@ -637,6 +333,311 @@
       this._conn.close();
     }
   };
+  MultiplexClient.prototype._parseMultiplexData = function(msg) {
+    try {
+      var m = /^(\d+)\|(m|o|c|r)\|([\s\S]*)$/m.exec(msg);
+      if (!m)
+        return null;
+      msg = {
+        id: m[1],
+        method: m[2],
+        payload: m[3]
+      };
+
+      switch (msg.method) {
+        case 'm':
+          break;
+        case 'o':
+          if (msg.payload.length === 0)
+            return null;
+          break;
+        case 'c':
+          try {
+            msg.payload = JSON.parse(msg.payload);
+          } catch(e) {
+            return null;
+          }
+          break;
+        case 'r':
+          break;
+        default:
+          return null;
+      }
+
+      return msg;
+
+    } catch(e) {
+      log('Error parsing multiplex data: ' + e);
+      return null;
+    }
+  };
+  // Open a new SockJS connection and assign it.
+  MultiplexClient.prototype._openConnection_p = function(){
+    var def = $.Deferred();
+
+    var url = this._sockjsUrl;
+    if (!this._first){
+      // Communicate to the server that we're intending to re-use an existing ID.
+      url = url.replace(/\/n=/, '/o=');
+    }
+    var conn = new SockJS(url,
+      null,{protocols_whitelist: this._whitelist});
+    var self = this;
+    conn.onmessage = function(){
+      self.onConnMessage.apply(self, arguments);
+    }
+
+    var self = this;
+
+    // Temporarily override so we can resolve the promise
+    conn.onopen = function(){
+      // Successful open; restore onClose pass through to real open callback
+      conn.onclose = function(){
+        self.onConnClose.apply(self, arguments);
+      };
+      self.onConnOpen.apply(self, arguments);
+      def.resolve(conn);
+    };
+    conn.onclose = function(err){
+      // If we got here, means we didn't get to onopen, so it failed.
+      def.reject(err);
+    };
+
+    this._conn = conn;
+
+    return def;
+  };
+  MultiplexClient.prototype.onConnMessage = function(e) {
+    var msg = this._parseMultiplexData(e.data);
```

Comment:
nit: `_parseMultiplexData` does not log its data in all cases when it returns `null`. We should either do that there or here at the call point.